_posts/2021-09-26-where-the-store-is-stored.md

```diff
@@ -0,0 +1,292 @@
+drilling](https://sebhastian.com/react-prop-drilling/). To overcome this, we
+can add a `Context` to make our state and its API available to every child
+components via a simple `useContext` - even if [it's not made for
+that](https://twitter.com/sebmarkbage/status/1219836431972978689). The idea is
+the same: the store is stored inside React "runtime" via `useState`.
+
+## Programatic store
+
+And finally, there's good old way: dependency injection by hand. Let's say we're
+in a full custom app (no React shenanigans) and we don't want to use modules. We
+can then craft a store instance when the app starts and pass it along everywhere
+via function parameters.
+
+Let's write a store factory very similar to the ES6 module option.
+
+```js
+// store.js
+function buildStore() {
+  const _cars = [];
+
+  return {
+    add: (car) => {
+      _cars.push(car);
+    },
+
+    list: () => {
+      return _cars;
+    },
+  }
+}
+```
+
+Now that our store's factory is ready, we just have to craft a store instance
+somewhere in our app's entrypoint and pass it along everywhere.
+
+```js
+// index.js
+
+function startApp(store) {
+  store.add('new car');
+
+  document.body.innerText = store.list().join(', ');
+
+  // ...
+}
+
+function initApp() {
+  const store = buildStore();
+
+  startApp(store);
+}
+
+initApp();
+```
+
+Of course, any function that needs to access the store will have to receive it
+via its parameters... We're right back where we left our props drilling issue.
+
+# Pros and Cons of each store location 
+
+OK so we've outlined three potential store locations. You can see me coming with
+the list of pros and cons for each of them. 
+
+Let's start with modules. Pros: it's raw ES6 (`import` / `export`) and it is
+quite simple to grasp. It's also very flexible. We can do more or less whatever
+we want in a module and every part of our codebase can use it.
+
+But there is a big constraint. Not all codebases always run in the same runtime.
+Actually a lot of single page applications run in both browser and NodeJS
+runtimes, where the contexts are very different. In the browser, we can make the
+assumption that there is only one user at a given time using the runtime. So our
+module state will be "our user state". On NodeJS, we will probably use the code
+to render the page server-side for multiple HTTP requests. In this situation,
+our module state will be shared between multiple. Boom! Enjoy the data leak
+security issue.
+
+For library or framework locations such as our React `useState`, the big
+advantage is probably the strong integration with the corresponding library.
+Even bigger store solutions like Redux come with their binding like
+`react-redux` that does the job for us. No need to handle the Context / Provider
+stuff I mentioned. No issue with server-side rendering: the store being "in
+React runtime", it will be scoped to the `ReactDOM.render` or
+`ReactDOMServer.renderToString` (which will be called once for every http
```

Comment:
As a rule of thumb, I always try to avoid parenthesis when writing. I think those ideas are better included as a plain sentence.